Label Support for Users

rsh52 opened the pull request:
# Description
This PR introduces the option for users to select between `raw` or `label` outputs from `read_redcap_tidy` which impacts multichoice field types (truefalse, yesno, checkboxes, radio buttons, and dropdown lists). The current default selection is `label`s.

To address this, I made a new function, `multi_choice_to_labels` (name can be changed), which is used in the overarching `read_redcap_tidy` function if `raw_or_label = "label"`. From there, the function `mutate`s yesno/truefalse/checkbox outputs to those specified in #8 using a `for` loop with a series of conditionals checked against the metadata `field_type`. 

Later in the loop, `parse_labels` is used on `radio` and `dropdown` `field_type`s to return label data. `factor`s are enforced and orders are specified in the order they are output by `parse_labels`. 

> **Note:** In the latter portion, I made use of `rlang` injections which I believe could be a little more slick. I could not get my `left_join(by = '')` to work well with this and made a comment about it.

# Proposed Changes 
List changes below in bullet format:

- Create function `multi_choice_to_labels`
- Create preliminary test suite to ensure `multi_choice_to_labels` returns expected data types
- Remove `raw_or_label` calls from functions where no longer necessary

### Issue Addressed
**Closes #8**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

skadauke commented on tests/testthat/test-clean_redcap.R:
```diff
@@ -45,8 +45,8 @@ test_that("extract_nonrepeat_table tibble contains expected columns and data typ
   expect_character(out$text)
```
I think you have to repeat this section on a data set in which `raw_or_label == label`.

skadauke commented on tests/testthat/test-clean_redcap.R:
```diff
@@ -45,8 +45,8 @@ test_that("extract_nonrepeat_table tibble contains expected columns and data typ
   expect_double(out$calculated)
-  expect_double(out$dropdown_single)
-  expect_double(out$radio_single)
+  expect_character(out$dropdown_single)
```
Assuming this is for `label` - shouldn't `dropdown_single` and `radio_single` be a `factor` rather than `character`?

rsh52 replied:
Yes, I'll look into this tomorrow.

rsh52 replied:
Actually I think this is correct here, I changed the content of the questions in the sample REDCap to have non-numeric labels and raw data (both sides of the comma), and `clean_redcap` does not enforce any label manipulation or cross-referencing with the metadata. That's currently done at the top level `read_redcap_tidy` separate from `clean_redcap`. So here I just needed to update what we expect to come out of REDCap itself, but reordering of things might be necessary.

skadauke commented on tests/testthat/test-clean_redcap.R:
```diff
@@ -45,8 +45,8 @@ test_that("extract_nonrepeat_table tibble contains expected columns and data typ
-  expect_double(out$radio_single)
+  expect_character(out$dropdown_single)
+  expect_character(out$radio_single)
   expect_double(out$checkbox_multiple___1)
```
Assuming this is for `label` - shouldn't `checkbox` and `truefalse` be `logical`? And `yesno` should be `factor`?

rsh52 replied:
Yes, I misread the issue and thought factor was only for the latter two (dropdown/radio). I can add an a quick coercion to the function that coerces to `factor`, I actually had it in an earlier commit and removed it.

rsh52 replied:
See above, these tests are based on `clean_redcap` and not the output of the new function. However, I did update the output of `yesno` in the newest commit.

skadauke reviewed (commented):
Hey - I'm not sure this is right yet, form looking at the test suite. Did you look at the output of running the classic test DB with `raw_or_label = label`? We can look at this tomorrow.

rsh52 commented:
> Hey - I'm not sure this is right yet, form looking at the test suite. Did you look at the output of running the classic test DB with `raw_or_label = label`? We can look at this tomorrow.

Yes, I have to re-address these tests. The behavior is working as expected though in the call below, but `yesno` needs to be fixed to be `logical` instead of `character`:

```r
read_redcap_tidy(redcap_uri = Sys.getenv("REDCAP_URI"),
                 token = Sys.getenv("REDCAPTIDIER_CLASSIC_API"),
                 raw_or_label = "label") %>% 
  bind_tables()
```

Where:

```r
> data_field_types$radio_single
[1] B    C    <NA>
Levels: A < B < C
```

```r
> data_field_types$dropdown_single
[1] one   three <NA> 
Levels: one < two < three
```

```r
> class(data_field_types$checkbox_multiple_2___3dddd)
[1] "logical"
```

```r
> class(data_field_types$truefalse)
[1] "logical"
```

rsh52 commented:
Hm, seems the R-CMD-Check is stuck and it won't let me cancel the workflow... Will check later in the day but it does work locally.

skadauke commented on R/utils.R:
```diff
@@ -150,3 +144,64 @@ update_field_names <- function(db_metadata, raw_or_label = 'raw'){
+
+multi_choice_to_labels <- function(db_data, db_metadata){
+
+  for (i in 1:nrow(db_metadata)) {
```
I would put the following ahead of the `for` loop and then delete lines 162-184.

```r
# Handle columns where we change 0/1 to FALSE/TRUE (logical)
logical_cols <- db_metadata %>%
  filter(field_type %in% c("yesno", "truefalse", "checkbox")) %>%
  pull(field_name_updated)
 
db_data2 <- db_data %>%
  mutate(across(.cols = logical_cols, as.logical))
```

rsh52 replied:
Just want to make sure, you mean `db-data <- db_data` not `db_data2 <- db_data` right? Accidentally blindly copy/pasted and had test fails because we aren't using `db_data2` elsewhere haha

skadauke commented on R/utils.R:
```diff
@@ -150,3 +144,64 @@ update_field_names <- function(db_metadata, raw_or_label = 'raw'){
+
+    # dropdown and radio datatype handling ----
+    if (db_metadata$field_type[i] == "dropdown" | db_metadata$field_type[i] == "radio") {
+
```
better: `db_metadata$field_type[i] %in% c("dropdown", "radio")`

skadauke commented on R/utils.R:
```diff
@@ -150,3 +144,64 @@ update_field_names <- function(db_metadata, raw_or_label = 'raw'){
+
+      # Replace values from db_data$(field_name) with label values from  parse_labels key
+      db_data <- db_data %>%
+        left_join(parse_labels_output %>% rename(!!field_name := raw), # Could not get working in by argument, instead inject field name for rename in parse_labels_output
```
I don't think the `rename` is necessary if you use a named `by` -- i.e. `by = c("raw" = field_name)` or similar

skadauke commented on R/utils.R:
```diff
@@ -150,3 +144,64 @@ update_field_names <- function(db_metadata, raw_or_label = 'raw'){
+
+      db_data <- db_data %>%
+        mutate(
+          across(.cols = field_name, .fns = ~factor(., ordered = T, levels = parse_labels_output$label))
```
I don't think we want to make `ordered = TRUE` the default.

rsh52 commented:
Changes implemented, checks passing (hooray!). Just need to resolve the new unrelated bug 😅

skadauke commented on tests/testthat/test-utils.R:
```diff
@@ -0,0 +1,28 @@
+  expect_logical(out$yesno)
+  expect_logical(out$truefalse)
+
+  checkbox_col_types <- map(out %>% select(contains("checkbox")), class)
```
Ugghhhh... can you make this less complicated? I really don't know what's going on here. There should really be no `for` loops or `map` functions in tests.

rsh52 replied:
Haha I knew you'd call this out. I was trying to make an elegant way of capturing all `checkbox_*` fields to test, but if you prefer we can either test one field or do many single tests. I think in the `clean_*` tests we just tested a single `checkbox_*` field.

skadauke commented on tests/testthat/test-utils.R:
```diff
@@ -0,0 +1,28 @@
+
+  checkbox_col_types <- map(out %>% select(contains("checkbox")), class)
+  checkbox_col_types <- data.frame(
+    matrix(unlist(checkbox_col_types),
```
???

rsh52 replied:
See above haha

skadauke reviewed (commented):
See comments about `test-utils.R`. Also, I think it would be good to address the bug(s) we discovered today inside of this PR.

rsh52 commented:
Updated `parse_labels` with the comments we made today over email and added a few tests for string expectations. The function will now look for the first comma in splits on `|`s, and will fail if:

- supplied matrix dimensions are not even
- split objects contain no commas

It will still fail silently though for the example you wrote in the red/green email:
`raw, label | continued label, with commas, and pipe | another, one ` 

Where it is both even and contains commas. Unsure how to tackle this one.

Also sadly calling Prodigy and CART still fails.. 😞  All of our test REDCap's pass though. Will have to look further into this on Monday.

rsh52 commented:
Brief update: The error related to CART/Prodigy turned into 4 errors, but I figured out and resolved 3 of them! Unfortunately I need to take the dog out and leave soon but I have all of the resolutions documented to propose when I get the last bug squashed. Since I'm running out the door, here are my jumbled notes:

> There is still one error related to Prodigy failing on raw/label matrix structure when evaluating labels.
 
## **Prodigy/CAR-T Error Resolution (label):**

When radio/dropdown values are assigned raw values of integers, REDCap outputs them as integers.
 
![image](https://user-images.githubusercontent.com/38384694/185710309-0089642d-bbae-44c5-8de9-4d8e98379c2c.png)

Solution is to coerce these values to character inside multi_choice_to_labels so left_join doesn’t fail.

## **Prodigy Error Resolution (raw):**
![image](https://user-images.githubusercontent.com/38384694/185710342-bb525fbc-edd5-49a2-9796-b1db4fd2ae3f.png)

Metadata exports were still including “descriptive” field_types, which don’t come out in the database. So email variables were failing because there was nothing to select on.
 
Quick fix is to just filter out all “descriptive” field types as soon as metadata is called.
 
## **CAR-T Export Error Resolution (raw):**
![image](https://user-images.githubusercontent.com/38384694/185710381-aa026ce6-ecbb-4516-b8be-6e912b394465.png)

This one is pretty funny, I was baffled why these weren’t coming out and they must be new additions to the database. Checked my user rights and saw that I did not have full dataset access to these instruments specifically:
![image](https://user-images.githubusercontent.com/38384694/185710403-f0e4dc89-05e5-4c8f-8f1c-8be7eeb7ac8d.png)
![image](https://user-images.githubusercontent.com/38384694/185710415-444b3202-0c5e-4936-affb-f7f6b3c80a00.png)

rsh52 commented:
Alright, found the last error was related to a descriptive text field in Prodigy with field embedding that was accidentally made as a radio field. The `parse_labels` failed because the `select_choices_*` variable was empty where a string was expected. I think we're good now! 🤞

skadauke commented on R/utils.R:
```diff
@@ -69,15 +69,31 @@ link_arms <- function(
+
+  # Check there is a comma in all | delimited strsplit elements
+  if (!all(grepl(",", out[[1]]))) {
+    stop(paste0("The raw/label structure of the string is not analyzable: ", string))
```
Do you want to use `checkmate::assert` here? 

Also, I would change the message to: "Cannot parse the select_choices_or_calculations field from REDCap metadata. This may happen if there is a pipe character \`|\` inside the label"

rsh52 replied:
I tried to initially but couldn't figure out how to get a custom error message out of `assert`/`check`. The default when I used it wasn't helpful for the user., i.e. `Error: Assertion on 'all(grepl(",", out[[1]]))' failed: Must be TRUE.`

I would still like to keep the `string` output because I think it can be helpful to the user to locate what field is causing the issue. Otherwise agree, implementing now.

skadauke commented on R/utils.R:
```diff
@@ -69,15 +69,31 @@ link_arms <- function(
+    unlist()
+
+  # Check if vector is even for matrix creation. If not, then fail.
+  if (length(out) %% 2 != 0) {
```
Same comment as above, would use the same error message

skadauke commented on tests/testthat/test-utils.R:
```diff
@@ -0,0 +1,42 @@
+  expect_equal(levels(out$radio_single), c("A", "B", "C"))
+})
+
+test_that("parse_labels works", {
```
This is my favorite test of all tests you've written! 🥇

rsh52 replied:
What high praise! Great way to start the weekend 🎉

skadauke reviewed (commented):
This looks great! 

I think we should gracefully deal with the situation when you have a parse_labels receives an empty string. It's not technically illegal in REDCap for a radio button or similar field to have no categories so I wouldn't cause failure here. But it's probably an error in the database so the sensible thing would be to throw a warning "Warning: The field {field_name} in {instrument_name} is a {field_type}, however it does not have any categories." I would add this into the current PR.

For the issue you encountered in the CART DB, I think we should make a separate Github issue. I think we should detect when not all of the tables or fields that we expect from the metadata are present. I think this should not fail, but there should be a warning that informs the analyst of which instruments are missing, and why that might be the case. It should be possible to suppress this warning as it may be intended if a database is set up to export only specific instruments and/or identifier fields are removed (what is the difference between "De-identified* and "Remove All Identifier Fields"??).

rsh52 commented:
> This looks great!
> 
> I think we should gracefully deal with the situation when you have a parse_labels receives an empty string. It's not technically illegal in REDCap for a radio button or similar field to have no categories so I wouldn't cause failure here. But it's probably an error in the database so the sensible thing would be to throw a warning "Warning: The field {field_name} in {instrument_name} is a {field_type}, however it does not have any categories." I would add this into the current PR.

🎉 🎉 🎉 

I agree, I can add this. In a way it's almost good I made some mistakes in REDCap 😆 

> For the issue you encountered in the CART DB, I think we should make a separate Github issue. I think we should detect when not all of the tables or fields that we expect from the metadata are present. I think this should not fail, but there should be a warning that informs the analyst of which instruments are missing, and why that might be the case. It should be possible to suppress this warning as it may be intended if a database is set up to export only specific instruments and/or identifier fields are removed (what is the difference between "De-identified* and "Remove All Identifier Fields"??).

Will get on writing this up, not sure how to make it not fail with the current setup. Unless we implement behavior that also removes this missing data from the metadata so there's no crash by the time `left_join` comes around? In that case I think a warning would be good.

IIRC I think de-identified/remove identifiers is a setting in REDCap that labels a variable as an identifier so would block it from export. It may also date-shift dates but I'm not sure, would have to test.

rsh52 commented:
@skadauke I created two more issues for `form_status_complete` and to address missing instrument handling (such as what I experienced with the API).

In the most recent commit here I added a `warning` check to the `multi_choice_to_labels` to check for empty `select_choices_or_calculations` strings using the message you provided. I also wrapped the existing error checks to pass if this is met (I made it a nested loop for easier reading). I also made two more tests to catch the behavior.

Let me know what you think and if we can go ahead and merge this.

skadauke reviewed (commented):
Looks great to me!
